Composite Primarry Keys Guide [WIP] (#18)
bhumi1102/rails: guides/source/active_record_composite_primary_keys.md

What are Composite Primary Keys?
--------------------------------

> Sometimes a single column's value isn't enough to uniquely identify every row of

Reviewer: I wonder if it's worth flipping this paragraph around. Starting with the word "sometimes" instantly puts the question "when?" in my mind. That question is answered further on, but I think a better structure would be to start with the problem cases and then state that composite primary keys are a solution to that problem.

Author: updated

a table, and a combination of two or more columns is required. This can occur
with legacy database schemas that lack a single `id` primary key, or in
applications where the schema has been designed to partition data across

> multiple databases (sharding) or isolate data per customer or tenant

Reviewer: ```suggestion
[multiple databases (sharding)](active_record_multiple_databases.html) or isolate data per customer or tenant
```

(multitenancy). Composite primary keys (CPK) are designed to solve this by allowing two or more columns to together act as the unique identifier for a row.

> Composite primary keys do increase complexity and can be slower than a

Reviewer: Might be worth changing this to a `NOTE`?

Author: done

single primary key column. Ensure your use case requires a composite primary key
before using one.

### Using `query_constraints` as an Alternative

In cases where your table has a conventional `id` column but you want Active Record to scope queries using an additional column, you can use [`query_constraints`](https://api.rubyonrails.org/classes/ActiveRecord/Persistence/ClassMethods.html#method-i-query_constraints) instead of redefining the primary key entirely. For example:

```ruby
class Developer < ActiveRecord::Base
  query_constraints :company_id, :id
end
```

> This keeps `id` as the primary key at the database level while instructing Active Record to always include `company_id` in queries, updates, and deletes. It's a lighter-weight option when you don't need a true composite primary key but want Rails to treat a combination of columns as the effective identity.

Reviewer: It might be an idea to demonstrate a basic SQL query with and without `query_constraints` so readers understand what's going on under the hood.

Author: done

Declaring Composite Primary Keys and Creating Migrations
--------------------------------------------------------

> To create a table with a composite primary key, you can pass an array to the `primary_key:` option in the migration file:

Reviewer: Nitpick:

```suggestion
Create a table with a composite primary key by passing an array to the `primary_key:` in the database migration:
```

Author: updated

      t.integer :store_id
      t.string :sku
      t.text :description
      t.timestamps
    end
  end
end
```

After running the migration above, your `schema.rb` file will reflect the composite key:

> ```ruby

Reviewer: I'm not entirely sure we need this snippet. The schema always reflects the current state of the DB so this is normal behaviour. I'd only document it if something was different about it.

Author: Yeah I can see that point. Leaning on keeping it, it's useful to see how the composite keys look in the schema.

# db/schema.rb
create_table "products", primary_key: [:store_id, :sku], force: :cascade do |t|
  t.integer "store_id", null: false
  t.string "sku", null: false
  t.text "description"
  t.datetime "created_at", null: false
  t.datetime "updated_at", null: false
end
```

> NOTE: When using a composite primary key, uniqueness is enforced by the combination of columns rather than a single auto-incrementing `id`. In the above example, `store_id` would typically be a foreign key to a `stores` table, and `sku` would be a string the application provides (like "ABC-123"). Neither needs to be auto-generated, their combination is what's unique. However, if your CPK contains no conventional `id` column, you are responsible for ensuring uniqueness, through application logic, UUIDs, etc.

Reviewer: `CPK` standing for composite primary key isn't obvious to me. Took me a second to figure it out ... I even typed out a comment asking what it stood for. If we're going to using this acronym in the guide, we need to introduce it at some point alongside the full form like: `Composite Primary Key (CPK)`.

Also if `store_id` being a foreign key is what's expected and best practice in this scenario, I think we should demonstrate that in the snippet with a `foreign_key` option.

Author: Updated it just spell out CPK. Since it's a mouthful tried the acronym but yeah if it leads to confusion probably not worth it. 

Added the foreign_key syntax.

> Rails also supports declaring composite primary keys at the model level via [`self.primary_key`](https://api.rubyonrails.org/classes/ActiveRecord/AttributeMethods/PrimaryKey/ClassMethods.html) method:

Reviewer: How is this different from doing it in a migration?

Author: This is a way to be explicit if needed, as explained below: In most cases, you don't need to declare a composite primary key with
`self.primary_key` in your model at all. If you define the composite primary key
in your migration and use the default `schema.rb` format, Rails will read the
primary key from the schema automatically at boot time and your model will work.

```ruby
class Order < ApplicationRecord
  self.primary_key = [:store_id, :sku]
end
```

This tells Active Record that records are uniquely identified by the combination of both columns, not a single `id`.

In most cases, you don't need to declare a composite primary key with
`self.primary_key` in your model at all. If you define the CPK in your migration
and use the default `schema.rb` format, Rails will read the primary key from the
schema automatically at boot time and your model will work.

On the other hand, if your application uses `structure.sql` instead of `schema.rb` or if you're connecting to a legacy or external database that was not created by your migrations, you can use `self.primary_key` to explicitly declare a composite primary key.

Querying Models
---------------

### Using `#find`

If your table uses a composite primary key, you'll need to pass an array

> when using `#find` to locate a record. For example to find the product with `store_id` 3 and `sku` "XYZ12345":

Reviewer: What will happen if you pass just the `id` to `find`?

Author: good question. added the answer to the text (ActiveRecord::RecordNotFound error)

end
```

For the association above, `Order` has a composite primary key of `[:store_id, :id]`, and `Book` belongs to `Order`. Rails infers that the foreign key column on the books table is `order_id`, and will use only the `id` portion of the composite key when querying the association.

When we create an `Order` and a `Book` associated with it:

```ruby
order = Order.create!(store_id: 1, id: 2, status: "pending")
book = order.books.create!(title: "A Cool Book")
book.order
```

Rails generates the following SQL to access the order:

```sql
SELECT * FROM orders WHERE id = 2
```

Rails uses only `id` here because the composite primary key does contain an `:id` column, and it is unique for all records, so the partial key is sufficient.

> Rails can often infer the primary key to foreign key relationships between associated models. However, when dealing with composite primary keys, Rails typically defaults to using only part of the composite key — usually the `id` column — unless explicitly instructed otherwise. This default behavior only works if the model's composite primary key contains the `:id` column, and that column is unique for all records.

Reviewer: This is a repeat of line 170 above, as are the next few lines.

Author: haha thanks for spotting. removed the duplicate copy/paste.

record in the view. The `extract_value` method may be used to extract arrays
out of any delimited parameters.

Fixtures
--------

> Fixtures for composite primary key tables are fairly similar to normal tables.

Reviewer: Do we need this line? I'm not a fan of terms like "fairly similar". They don't add any value in my mind, just add fluff.

Author: updated
